include/pybind11/pybind11.h

```diff
@@ -802,12 +802,52 @@ class generic_type : public object {
 
     static void releasebuffer(PyObject *, Py_buffer *view) { delete (buffer_info *) view->internal; }
 };
+
+template <template<typename> class Predicate, typename... BaseTypes> struct class_selector;
+template <template<typename> class Predicate, typename Base, typename... Bases>
+struct class_selector<Predicate, Base, Bases...> {
+    static inline void set_bases(detail::type_record &record) {
+        if (Predicate<Base>::value) record.base_type = &typeid(Base);
+        else class_selector<Predicate, Bases...>::set_bases(record);
+    }
+};
+template <template<typename> class Predicate>
+struct class_selector<Predicate> {
+    static inline void set_bases(detail::type_record &) {}
+};
+
 NAMESPACE_END(detail)
 
-template <typename type, typename holder_type = std::unique_ptr<type>, typename type_alias = type>
+template <typename type_, typename... options>
 class class_ : public detail::generic_type {
+    template <typename T> using is_holder = detail::is_holder_type<type_, T>;
+    template <typename T> using is_subtype = detail::bool_constant<std::is_base_of<type_, T>::value && !std::is_same<T, type_>::value>;
+    template <typename T> using is_base_class = detail::bool_constant<std::is_base_of<T, type_>::value && !std::is_same<T, type_>::value>;
+    template <typename T> using is_valid_class_option =
+        detail::bool_constant<
+            is_holder<T>::value ||
+            is_subtype<T>::value ||
+            is_base_class<T>::value
+        >;
+
+    using extracted_holder_t = typename detail::first_of_t<is_holder, options...>;
+
 public:
-    typedef detail::instance<type, holder_type> instance_type;
+    using type = type_;
+    using type_alias = detail::first_of_t<is_subtype, options...>;
+    constexpr static bool has_alias = !std::is_void<type_alias>::value;
+    using holder_type = typename std::conditional<
+        std::is_void<extracted_holder_t>::value,
+        std::unique_ptr<type>,
+        extracted_holder_t
+    >::type;
```

Comment:
Ah right, sorry, I didn't see that `has_alias` is used in multiple places. Definitely keep it as `constexpr`, it's more convenient since it's only used as a value.